src/states/establishing_node.rs

```diff
@@ -0,0 +1,463 @@
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{
+    common::{
+        proxied, Approved, Base, Bootstrapped, NotEstablished, Relocated, RelocatedNotEstablished,
+    },
+    node::Node,
+};
+use crate::{
+    ack_manager::AckManager,
+    cache::Cache,
+    chain::{Chain, GenesisPfxInfo, ProvingSection, SectionInfo},
+    error::RoutingError,
+    id::{FullId, PublicId},
+    messages::{DirectMessage, HopMessage, Message, RoutingMessage},
+    outbox::EventBox,
+    parsec::{self, Block, GossipMessage, Parsec},
+    peer_manager::{Peer, PeerManager, PeerState},
+    routing_message_filter::RoutingMessageFilter,
+    routing_table::{Authority, Prefix},
+    state_machine::{State, Transition},
+    time::{Duration, Instant},
+    timer::Timer,
+    xor_name::XorName,
+    Service,
+};
+use itertools::Itertools;
+use std::{
+    collections::BTreeSet,
+    fmt::{self, Display, Formatter},
+};
+
+const POKE_TIMEOUT: Duration = Duration::from_secs(60);
+
+pub struct EstablishingNode {
+    ack_mgr: AckManager,
+    cache: Box<Cache>,
+    chain: Chain,
+    crust_service: Service,
+    full_id: FullId,
+    gen_pfx_info: GenesisPfxInfo,
+    /// Routing messages addressed to us that we cannot handle until we are established.
+    msg_backlog: Vec<RoutingMessage>,
+    notified_nodes: BTreeSet<PublicId>,
+    parsec: Parsec,
+    peer_mgr: PeerManager,
+    poke_timer_token: u64,
+    routing_msg_filter: RoutingMessageFilter,
+    timer: Timer,
+}
+
+impl EstablishingNode {
+    #[allow(clippy::too_many_arguments)]
+    pub fn from_proving_node(
+        ack_mgr: AckManager,
+        cache: Box<Cache>,
+        crust_service: Service,
+        full_id: FullId,
+        gen_pfx_info: GenesisPfxInfo,
+        min_section_size: usize,
+        msg_backlog: Vec<RoutingMessage>,
+        notified_nodes: BTreeSet<PublicId>,
+        peer_mgr: PeerManager,
+        routing_msg_filter: RoutingMessageFilter,
+        timer: Timer,
+    ) -> Self {
+        let public_id = *full_id.public_id();
+        let poke_timer_token = timer.schedule(POKE_TIMEOUT);
+
+        let parsec = parsec::create(full_id.clone(), &gen_pfx_info);
+        let chain = Chain::new(min_section_size, public_id, gen_pfx_info.clone());
+
+        let node = Self {
+            ack_mgr,
+            cache,
+            chain,
+            crust_service,
+            full_id,
+            gen_pfx_info,
+            msg_backlog,
+            notified_nodes,
+            parsec,
+            peer_mgr,
+            poke_timer_token,
+            routing_msg_filter,
+            timer,
+        };
+
+        debug!("{} - State changed to EstablishingNode.", node);
+        node
+    }
+
+    pub fn into_node(
+        self,
+        sec_info: SectionInfo,
+        old_pfx: Prefix<XorName>,
+        outbox: &mut EventBox,
+    ) -> State {
+        let node = Node::from_establishing_node(
```

Comment:
This is not ideal since it rely on type being different to ensure things are ok. And I'm not sure we were doing the right thing with message_backlog comment above.
I'm wondering if we should have a struct here that we pass in from_establishing_node, would also avoid #[allow(clippy::too_many_arguments)]. (Not in this commit either follow up or preceding).

Reply:
We can add a struct to avoid the `too_many_arguments` lint, but then we would be just adding more code to satisfy the linter while not really making things cleaner. Seems pointless to me. On the other hand, we could change `Node::from_establishing_node` to take `EstablishingNode` directly, but then we would have to change all fields in `EstablishingNode` to `pub(super)` or something. Might be worth it, not sure.

Reply:
The benefit i see is with a struct, the field names have to be specified.
I do not think we want to the API with what fields we need in the EstablishingNode iteself.
```
function(
      value1,
      value2,
);

Struct {
     value1,
     value3: value2
};
```

Reply:
It's not public API though.

Reply:
`pub fn from_proving_node(`.
In my mind it it the public API of Node with respect to the other mod, even if it is not use outside the crate.

Reply:
OK. This might be worth doing, but this pattern is used in other states too, so best to do it everywhere to maintain consistency.